pkg/auth/providers/oidc/oidc_provider.go

```diff
@@ -0,0 +1,389 @@
+	client "github.com/rancher/rancher/pkg/client/generated/management/v3"
+	publicclient "github.com/rancher/rancher/pkg/client/generated/management/v3public"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/rancher/pkg/user"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	Name      = "oidc"
+	UserType  = "user"
+	GroupType = "group"
+)
+
+type Provider struct {
+	CTX         context.Context
+	AuthConfigs v3.AuthConfigInterface
+	Secrets     corev1.SecretInterface
+	UserMGR     user.Manager
+	TokenMGR    *tokens.Manager
+}
+
+type ClaimInfo struct {
+	Subject           string   `json:"sub"`
+	Name              string   `json:"name"`
+	PreferredUsername string   `json:preferred_username`
+	GivenName         string   `json:given_name`
+	FamilyName        string   `json:family_name`
+	Email             string   `json:"email"`
+	EmailVerified     bool     `json:"email_verified"`
+	Groups            []string `json:"groups"`
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager, tokenMGR *tokens.Manager) common.AuthProvider {
+	return &Provider{
+		CTX:         ctx,
+		AuthConfigs: mgmtCtx.Management.AuthConfigs(""),
+		Secrets:     mgmtCtx.Core.Secrets(""),
+		UserMGR:     userMGR,
+		TokenMGR:    tokenMGR,
+	}
+}
+
+func (o *Provider) GetName() string {
+	return Name
+}
+
+func (o *Provider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = o.ActionHandler
+	schema.Formatter = o.Formatter
+}
+
+func (o *Provider) AuthenticateUser(ctx context.Context, input interface{}) (v3.Principal, []v3.Principal, string, error) {
+	login, ok := input.(*v32.OIDCLogin)
+	if !ok {
+		return v3.Principal{}, nil, "", fmt.Errorf("unexpected input type")
+	}
+	return o.LoginUser(ctx, login, nil)
+}
+
+func (o *Provider) LoginUser(ctx context.Context, oauthLoginInfo *v32.OIDCLogin, config *v32.OIDCConfig) (v3.Principal, []v3.Principal, string, error) {
+	var userPrincipal v3.Principal
+	var groupPrincipals []v3.Principal
+	var claimInfo ClaimInfo
+	var err error
+
+	if config == nil {
+		config, err = o.GetOIDCConfig()
+		if err != nil {
+			return userPrincipal, nil, "", err
+		}
+	}
+
+	if config.Certificate != "" && config.PrivateKey != "" {
+		cert, err := tls.X509KeyPair([]byte(config.Certificate), []byte(config.PrivateKey))
+		if err != nil {
+			return userPrincipal, groupPrincipals, "", err
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM([]byte(config.Certificate))
+		oidcClient := &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					Certificates: []tls.Certificate{cert},
+					RootCAs:      caCertPool,
+				},
+			},
+		}
+		ctx = oidc.ClientContext(ctx, oidcClient)
```

Comment:
Should this be handled for each connection? Maybe need a createConnection utility method to set all the properties every time?

Reply:
Added this as a common util so other providers could use it if they needed.